cluster-autoscaler/cloudprovider/gce/mig_info_provider.go

```diff
@@ -151,13 +151,101 @@ func (c *cachingMigInfoProvider) getCachedMigForInstance(instanceRef GceRef) (Mi
 func (c *cachingMigInfoProvider) RegenerateMigInstancesCache() error {
 	c.cache.InvalidateAllMigInstances()
 	c.cache.InvalidateAllInstancesToMig()
-	migs := c.migLister.GetMigs()
-	errors := make([]error, len(migs))
-	workqueue.ParallelizeUntil(context.Background(), c.concurrentGceRefreshes, len(migs), func(piece int) {
-		errors[piece] = c.fillMigInstances(migs[piece].GceRef())
-	}, workqueue.WithChunkSize(c.concurrentGceRefreshes))
+	c.cache.InvalidateMigInstancesState()
+	err := c.fillMigInfoCache()
+	if err != nil {
+		return err
+	}
+
+	instances, err := c.listInstancesInAllZonesWithMigs()
+	if err != nil {
+		return err
+	}
+	migToInstances := groupInstancesToMigs(instances)
+	err = c.updateMigInstancesCache(migToInstances)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *cachingMigInfoProvider) listInstancesInAllZonesWithMigs() ([]GceInstance, error) {
+	zones := c.listAllZonesWithMigs()
+	var instances []GceInstance
+	for zone := range zones {
+		instancesOfZone, err := c.gceClient.FetchAllInstances(c.projectId, zone, "")
+		if err != nil {
+			return instances, err
+		}
+		instances = append(instances, instancesOfZone...)
+	}
+	return instances, nil
+}
+
+func groupInstancesToMigs(instances []GceInstance) map[GceRef][]GceInstance {
+	migToInstances := map[GceRef][]GceInstance{}
+	for _, instance := range instances {
+		if _, found := migToInstances[instance.Igm]; found {
```

Comment:
igm can be an empty GceRef, or a non-existent GceRef how do we want to handle that? https://github.com/kubernetes/autoscaler/blob/6d54cb73bcb0fad988985441b67006a017c1adbc/cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go#L412

Reply:
We update cache only for the migs that are known and simply ignore instances from empty GceRefs or non-existent GceRefs. Did you notice any issue with that?